src/Lecture4.hs

```diff
@@ -134,8 +138,33 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+splitRow :: String -> [String]
+splitRow = wordsWhen (==',')
+  where
+    wordsWhen     :: (Char -> Bool) -> String -> [String]
+    wordsWhen p s =  case dropWhile p s of
+                      "" -> []
+                      s' -> w : wordsWhen p s''
+                            where (w, s'') = break p s'
+
+readProduct :: String -> Maybe String
+readProduct "" = Nothing
+readProduct p = Just p
+
+readTrade :: String -> Maybe TradeType
+readTrade "Buy"  = Just Buy
+readTrade "Sell" = Just Sell
+readTrade _      = Nothing
+
+readCost :: String -> Maybe Int
+readCost = readMaybe
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow = readRow . splitRow
+   where readRow :: [String] -> Maybe Row
+         readRow cols = case cols of
+             [c1, c2, c3] -> Row <$> readProduct c1 <*> readTrade c2 <*> readCost c3
+             _ -> Nothing
```

Comment:
You have an excellent Applicative implementation 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Applying this programming pattern and a few helper functions, the `parseRow` function will look like this:

```suggestion
parseRow row = do
    [c1, c2, c3] <- Just $ splitRow row
    guard $ c1 /= ""
    trade <- readMaybe c2
    cost <- readMaybe c3
    guard $ cost >= 0
    pure $ Row c1 trade cost
```